cluster/provision_minion.sh

```diff
@@ -289,70 +404,450 @@ function stop_kube_svcs()
+        # Setup virtual-gateway
+        sed -i 's/# routing_instance=default-domain:admin:public:public$/routing_instance=default-domain:default-project:Public:Public/g' $vrac
+        sed -i 's/# interface=vgw$/interface=vgw/g' $vrac
+        PUBLIC_IP=$(echo $OPENCONTRAIL_PUBLIC_SUBNET | cut -d '/' -f 1)
+        PUBLIC_LEN=$(echo $OPENCONTRAIL_PUBLIC_SUBNET | cut -d '/' -f 2)
+        vgwipblk="'$PUBLIC_IP'\/'$PUBLIC_LEN'"
+        sed -i 's/# ip_blocks=1\.1\.1\.1\/24$/ip_blocks='$PUBLIC_IP'\/'$PUBLIC_LEN'/g' $vrac
+        sed -i "/GATEWAY-0/a routes=0.0.0.0/0" $vrac
+    fi
+}
+
+function persist_hostname()
+{
+   if [ ! -f /etc/hostname ]; then
+     echo "$hname" > /etc/hostname
+     hostname $hname
+   fi
+}
+
+function rpf_disable()
+{
+ echo 0 > /proc/sys/net/ipv4/conf/all/rp_filter
+ echo 0 > /proc/sys/net/ipv4/conf/$OPENCONTRAIL_VROUTER_INTF/rp_filter
+ echo 0 > /proc/sys/net/ipv4/conf/$VHOST/rp_filter
+ echo "net.ipv4.conf.all.rp_filter=0" >> /etc/sysctl.conf
+ echo "net.ipv4.conf."$OPENCONTRAIL_VROUTER_INTF".rp_filter=0" >> /etc/sysctl.conf
+ echo "net.ipv4.conf."$VHOST".rp_filter=0" >> /etc/sysctl.conf
+}
+
+function add_static_route()
+{
+  itf="/etc/network/interfaces"
+  if isGceVM ; then
+     # Assuming the the default kubernetes container used is debain based
+     # works only for debian / ubuntu
+     ocpubgw=$(sipcalc $OPENCONTRAIL_PUBLIC_SUBNET | grep "Usable range" | awk '{print $6}')
+     ocpubgwname=$(echo $ocpubgw | sed 's/\./-/g')
+     ocpubmask=$(sipcalc $OPENCONTRAIL_PUBLIC_SUBNET | grep "Network mask" | head -n 1 | awk '{print $4}')
+     zone=$(gcloud compute instances list | grep minion -A 1 | awk '{print $2}')
+     srt=$(gcloud compute routes list | grep -ow ip-$ocpubgwname)
+     if [ ! -z "$srt" ]; then
+         gcloud compute routes delete ip-$ocpubgwname <<< "y
+y
+y
+"
+     fi
+     gcloud compute routes create ip-$ocpubgwname --next-hop-instance `hostname` --next-hop-instance-zone $zone --destination-range $OPENCONTRAIL_PUBLIC_SUBNET
+     # create and configure vgw with static IP
+     # vgw will act as the gateway for public subnet
+     grep -q 'auto vgw' $itf || echo -e "\nauto vgw" >> $itf
+     grep -q 'iface vgw inet static' $itf || echo "iface vgw inet static" >> $itf
+     grep -q "address $ocpubgw" $itf || echo -e "    address $ocpubgw\n    netmask $ocpubmask" >> $itf
+     /etc/init.d/networking restart
+  fi
+  # Since we have vgw plumbed by both gke and vrouter there will
+  # 2 routes for the same prefix. It will not have any issue though.
+  # just for keeping it clean, delete the duplicate route
+  net=$(echo $OPENCONTRAIL_PUBLIC_SUBNET | cut -d "/" -f 1)
+  dups=$(route -n | grep vgw | grep $net | wc -l)
+  if [ $dups -gt 1 ]; then
+    route delete -net $OPENCONTRAIL_PUBLIC_SUBNET dev vgw
+  fi
+}
 
 function main()
```

Comment:
We need to break down this shell scripts...
There are common parts between gateway and minion.

Roughly:
1. Install the required binaries (NOP if the binaries are installed)
2. Configure the vhost0 interface  (NOP if the interface is already running).
3. Generate configuration files (this needs to be idempotent).
4. Start the agent.